#!/bin/bash
#PBS -q condo
#PBS -N chipseq
#PBS -l nodes=1:ppn=8
#PBS -l walltime=08:00:00
#PBS -o /home/zhc268/logs/chipseq.out
#PBS -e /home/zhc268/logs/chipseq.err
#PBS -V
#PBS -M zhangc518@gmail.com
#PBS -m abe
#PBS -A epigen-group

## RUN chipseq using two steps
# Step 1: process tr & ctl equally before peak calling
# Step 2: peak calling 

source activate aquas_chipseq



# select libs 
samplenames=(`cat $samples`)
INPREFIX=${samplenames[${PBS_ARRAYID}*3]} #index start from 0
genome=${samplenames[${PBS_ARRAYID}*3+1]}
ctl=${samplenames[${PBS_ARRAYID}*3+2]} #control or treatment
fastq1="${FASTQDIR}/${INPREFIX}_R1.fastq.gz"
fastq2="${FASTQDIR}/${INPREFIX}_R2.fastq.gz"


# parameters
type="histone"
true_rep="true"
no_pseudo_rep="true"
outdir="${WORKDIR}${INPREFIX}_chip"; mkdir -p $outdir

############################################################
# Step 1
############################################################
fastq_input="-fastq1_1 ${fastq1} -fastq1_2 ${fastq2}"
[[ "$ctl" = true ]] && fastq_input="-ctl_fastq1_1 ${fastq1} -ctl_fastq1_2 ${fastq2}"

#final_stage="filt_bam"
final_stage="xcor"

# run pipeline
bds /projects/ps-epigen/software/chipseq_pipeline/chipseq.bds \
    -type $type \
    -final_stage $final_stage \
    -true_rep $true_rep \
    -no_pseudo_rep $no_pseudo_rep \
    -nth $PBS_NP \
    -out_dir $outdir \
    -species $genome \
    $fastq_input 

wait 

############################################################
# Step 2
############################################################
[[ "$ctl" = true ]] && exit 0 # no step 2 for ctl 
[[ ! -z ${no_peak+x} ]] &&  exit 0 # no step 2 if -v no_peak

ctl_id=$(cat $samples | grep true -n |cut -d':' -f1)
ctl_id=$[$ctl_id-1] #0-based index
ctl_name=${samplenames[${ctl_id}*3]}
input_dir="${WORKDIR}${ctl_name}_chip/align"; 
ctl_prefix="$input_dir/ctl1/${samplenames[${ctl_id}*3]}" #index start from 0
tr_prefix="$outdir/align/rep1/${samplenames[${PBS_ARRAYID}*3]}" #index start from 0
#tag_suffix="_R1.PE2SE.nodup.tagAlign.gz"
tag_suffix="_R1.PE2SE.nodup.bam"
#tag_input="-tag ${tr_prefix}${tag_suffix} -ctl_tag ${ctl_prefix}${tag_suffix}"
tag_input="-filt_bam ${tr_prefix}${tag_suffix} -ctl_filt_bam ${ctl_prefix}${tag_suffix}"

# run pipeline
bds /projects/ps-epigen/software/chipseq_pipeline/chipseq.bds \
    -type $type \
    -true_rep $true_rep \
    -no_pseudo_rep $no_pseudo_rep \
    -nth $PBS_NP \
    -out_dir $outdir \
    -species $genome \
    -pe \
    $tag_input 

wait 

############################################################
# runFastQC & fastq_screen
############################################################

#runFastQC_screen.sh $INPREFIX
#results_transfer.sh $INPREFIX


#eg: qsub  -t 0-2 -v samples="$(pwd)/samples.txt",FASTQDIR=$(pwd),WORKDIR="/home/zhc268/scratch/outputs/"  ./runPipeline_chipseq_twoSteps.pbs 

#eg:qsub -t 0-3 -v samples="/home/zhc268/data/logs/zhc268/run_2018-03-01_AVD_chip.txt",FASTQDIR="/home/zhc268/data/seqdata/",WORKDIR="/home/zhc268/scratch/outputs/" ./runPipeline_chipseq_twoSteps.pbs
